File: cpp/p35.cpp
#include "euler.h"

using namespace std;

string rotate( string s ) {
  char c = s[ 0 ];
  s.erase( 0, 1 );
  s.push_back( c );

  return s;
}

bool is_circular_prime( int n ) {
  string n_string = number_to_string( n );

  for( int i = 0; i < n_string.size(); i++ ) {	
    if( !is_prime( string_to_number( n_string ) ) ) 
      return false;

    n_string = rotate( n_string );
  }
	
  return true;
}

int main() {
  int count = 0;

  for( int n = 0; n < 1000000; n++ )
    if( is_circular_prime( n ) ) 
      count++;
	
  cout << count << endl;
}

File: clojure/p35.clj
(ns p35
  (:use [euler :only (digits prime? prime-sieve timed-test)]))

(defn rotate [s]
  (str (subs s 1) (subs s 0 1)))

(defn rotations [n]
  (let [n-str (str n)
        strs (take (count n-str) (iterate rotate n-str))]
    (map #(Integer/parseInt %) strs)))

(timed-test
 55
 (count (reduce (fn [circular-primes prime]
                  (let [rots (rotations prime)]
                    (if (and (not (circular-primes prime))
                             (every? odd? (digits prime))
                             (every? prime? rots))
                      (apply conj circular-primes rots)
                      circular-primes)))
                #{2}
                (prime-sieve 1000000))))
